### PR TITLE
Alerting: Support for re-ordering alert rules in a group

### LINKED ISSUE #51494: Alerting: Update UI to let user manage order of the rules in a group
See https://github.com/grafana/grafana/issues/48353

### PR DESCRIPTION
**Pull request by gillesdemey**
**What this PR does / why we need it**:

Adds support for re-ordering rules in a rule group by navigating to the "grouped" view in the alert list.

https://user-images.githubusercontent.com/868844/182892722-6305a213-6426-486f-b8cd-e6877df945a7.mp4

**Which issue(s) this PR fixes**:

Fixes #51494 

**Special notes for your reviewer**:

Had to refactor the existing `hashRulerRule` function to return strings instead of numbers but I don't believe this impacts functionality.

### CODE REVIEW

File: public/app/features/alerting/unified/components/rules/ReorderRuleGroupModal.tsx [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+      if (sameIndex) {
+        return;
+      }
+
+      const newOrderedRules = reorder(rulesList, result.source.index, result.destination.index);
+      setRulesList(newOrderedRules); // optimistically update the new rules list
+
+      // the rulesSource is sometimes a string (for Grafana managed) or a rulesSource object
+      const rulesSourceName =
+        typeof namespace.rulesSource === 'string' ? namespace.rulesSource : namespace.rulesSource.name;
```
Comment (konrad147) [inline public/app/features/alerting/unified/components/rules/ReorderRuleGroupModal.tsx:58]: You can use the `getRulesSourceName` function which does the same

File: public/app/features/alerting/unified/components/rules/ReorderRuleGroupModal.tsx [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+      const rulerRules = compact(newOrderedRules.map((rule) => rule.rulerRule));
```
Comment (konrad147) [inline public/app/features/alerting/unified/components/rules/ReorderRuleGroupModal.tsx:60]: Why `compact` is needed?

**Review by konrad147 — APPROVED** [review]
Nice! Only minor nits

File: public/app/features/alerting/unified/components/rules/ReorderRuleGroupModal.tsx (hunk 2 above)
Comment (gillesdemey) [reply public/app/features/alerting/unified/components/rules/ReorderRuleGroupModal.tsx:60]: Great question! 

According to the type definition the `rulerRule` is an optional property, so the `rulerRules` variable is typed as `(CombinedRule | undefined)[]` – this conflicts with what `updateRulesOrder` wants (which is `CombinedRule[]`).

If we use the `compact` function we can omit any rules without `rulerRule` and satisfy the type checker at the same time.

I couldn't get it to work with `[].filter().map()` either and didn't want to use a type cast either.

File: public/app/features/alerting/unified/utils/rule-id.ts [hunk 3]
```diff
@@ -173,10 +173,10 @@ function hash(value: string): number {
   return hash;
 }
 
-// this is used to identify lotex rules, as they do not have a unique identifier
-function hashRulerRule(rule: RulerRuleDTO): number {
+// this is used to identify rules, mimir / loki rules do not have a unique identifier
+export function hashRulerRule(rule: RulerRuleDTO): string {
```
Comment (peterholmberg) [inline public/app/features/alerting/unified/utils/rule-id.ts:177]: do we want a test for this function?

**Review by peterholmberg — APPROVED** [review]
Looks good, works good. Have two comments :)

File: public/app/features/alerting/unified/utils/rule-id.ts (hunk 3 above)
Comment (gillesdemey) [reply public/app/features/alerting/unified/utils/rule-id.ts:177]: Yea, I think we most certainly do – will append one

**Pull request merged**